### ISSUE TITLE
Feature request: amp-list to support masonry layout

### ISSUE THREAD

**Issue by leonalicious** [ampproject/amphtml#19866]
Good Day,  here is some little background, in order to **filter** those unnecessary info from user sights & increase the **exposure** of important information, we are hoping is it possible we can have a **Horizontal** `masonry` layout option for `amp-list` please? 

Based on my research, there are 2 ways of getting this layout on the ground:

1. PURELY CSS solution with the use of `flex`, we can achieve the `vertical` layout easily compare to `horizontal`, but in real world, most people will read the contents from `LEFT to RIGHT` and from `TOP to BOTTOM`, that's why, `Horizontal` layout makes more sense here.
![image](https://user-images.githubusercontent.com/4065175/49936654-7f4ac900-ff0f-11e8-97de-280c8a95b4ff.png)
[source taken from : [hackernoon](https://hackernoon.com/masonry-layout-technique-react-demo-of-100-css-control-of-the-view-e4190fa4296)]

But with a little trick on it, we can get what we want by writing some `ugly code` which renders 2 sets of list , and after that, use `CSS nth-child` to hide the ODD and EVEN based on needs.

![ugly code to render 2 lists](https://user-images.githubusercontent.com/4065175/49937987-6d6b2500-ff13-11e8-90b1-27463e3bfd24.jpg)

OR apply the similar `ugly code` to re-arrange the sequence of items at our API endpoint.

2. Uses `absolute` position and some calculation as what [AMP version of Pinterest is doing](https://www.pinterest.com/explore/highland-cattle/).

In short term, we will go for the 1st approach but moving forward, hope this layout can be supported within `amp-list` or new component `amp-masonry`. This masonry layout feature will definitely make a perfect combination with the `infinite scroll` of what @cathyxz is currently working on, anyway, any thoughts ?

**Comment by cathyxz** [comment]
@alanorozco once wrote up a doc for a generic `amp-masonry` approach using [Houdini's Layout API](https://github.com/w3c/css-houdini-drafts/blob/master/css-layout-api/EXPLAINER.md), (code example [here](https://github.com/GoogleChromeLabs/houdini-samples/tree/master/layout-worklet/masonry) ) however [support](https://ishoudinireadyyet.com/) for this API isn't ready yet. I think the last discussed status of this feature was that we'd revisit it in a year when there is better support for Houdini's Layout API. 

We already support inline styles per https://github.com/ampproject/amphtml/pull/16544, which makes `approach 2` that you mentioned above somewhat feasible with infinite scroll.

**Comment by nainar** [comment]
Hi @leonalicious would doing this with `<amp-script>` be a possibility?

**Comment by leonalicious** [comment]
Good Day @nainar, I think so, but not too sure about the performance,  but I am sure with `handlebar` template, it would be a more perfect solution to support this as the key here is to be able to differentiate the data based on `odd` or `even`, how about your thoughts on how `amp-script` can benefit this ?

**Comment by leonalicious** [comment]
Today I have tried to implement `masonry` layout, but too bad, not that smooth as `<div role="listitem">` will be created and appended into the root of `amp-list` and cause children nodes of items are relying to different `div wrapper`

Understand the logic that `<div role="listitem">` will be created only IF
1. during the `mustache` rendering, there is more than one elements are being rendered, amp-list will create `<div role="listitem">` to wrap them, just like our current [search List page](https://m.aliexpress.com/wholesale/dress.html), wrapped `List <a>` &  `Gallery <a>` index by index throughout the Loop of mustache. 

![image](https://user-images.githubusercontent.com/4065175/54306171-bfe7da80-4603-11e9-87c2-a8edb6ab50b0.png)

With this approach, because each items are wrapped Row by Row, no chance i can achieve masonry which is Column by Column design.

2. With the option `single-items` on,  amp-list will append `<div role="listitem">` as well

not too sure how can I bypass and make this`<div role="listitem">` not to be created, any thoughts ?

**Comment by cathyxz** [comment]
Hi @leonalicious: 
So here's a simplistic demo for the three-list approach @sparhami proposed today: https://amp-infinite-scroll.glitch.me/. 
You can play with the source code at: https://glitch.com/edit/#!/amp-infinite-scroll
Click on the "Remix Project" button to get an editable version. 
<img width="291" alt="Screenshot 2019-04-17 at 5 15 42 PM" src="https://user-images.githubusercontent.com/1528181/56271889-76c00480-6134-11e9-8ecb-380b5ae9c766.png">

You need to make a few more server-side changes to get varied heights, but I think that approach actually generally works.

**Comment by leonalicious** [comment]
hey @cathyxz , tried with your solution, although it looks weird to me, it works fine with the `position: absolute` & `layout: fixed-height`. 

btw, I found another issue with the implementation, the request seems has gone hotwired even though I have stopped scrolling. Please refer to the gif below, thank you ~

![ezgif-3-5632a3c1ce98](https://user-images.githubusercontent.com/4065175/56504354-d78f7880-654a-11e9-8f36-737dbdea9d1e.gif)

**Comment by cathyxz** [comment]
Yeah I suspect that's a different and actual issue with `load-more`. I'm trying to find a cleaner repro to debug.

**Comment by leonalicious** [comment]
hey hi @cathyxz, are you able to reproduce this issue and fix it ?

**Comment by nainar** [comment]
Hi @leonalicious is the most recent issue one with infinite scroll or with masonry layout. If infinite scroll can you split this out into a separate issue and we can get someone to work on it?

**Comment by andreyMakahov** [comment]
Hi, @nainar I will continue working with masonry layout. Let me do my investigation first. After I will let you know what the issues we are facing.
